databases/tests/test_transactions.py

```diff
@@ -212,3 +212,129 @@ async def test_transaction_already_closed(client: Prisma) -> None:
         await transaction.user.delete_many()
 
     assert exc.match('Transaction already closed')
+
+
+@pytest.mark.asyncio
+@pytest.mark.skipif(CURRENT_DATABASE in ['cockroachdb', 'sqlite'], reason='Not available')
+async def test_read_uncommited_isolation_level(client: Prisma) -> None:
+    """A transaction isolation level is set to `READ_UNCOMMITED`"""
+    client2 = Prisma()
+    await client2.connect()
+
+    user = await client.user.create(data={'name': 'Robert'})
+
+    async with client.tx(isolation_level=prisma.TransactionIsolationLevel.READ_UNCOMMITED) as tx1:
+        tx1_user = await tx1.user.find_first_or_raise(where={'id': user.id})
+        tx1_count = await tx1.user.count()
+
+        async with client2.tx() as tx2:
+            await tx2.user.update(data={'name': 'Tegan'}, where={'id': user.id})
+            await tx2.user.create(data={'name': 'Bobby'})
+
+            dirty_user = await tx1.user.find_first_or_raise(where={'id': user.id})
+
+        non_repeatable_user = await tx1.user.find_first_or_raise(where={'id': user.id})
+        phantom_count = await tx1.user.count()
+
+    # Have dirty read
+    assert tx1_user.name != dirty_user.name
+    # Have non-repeatable read
+    assert tx1_user.name != non_repeatable_user.name
+    # Have phantom read
+    assert tx1_count != phantom_count
```

Comment:
@RobertCraigie Hi! I tried to add transaction isolation levels feature, but i don't understant, how to test it (tests not passes). I will be very grateful, if you help me find my mistake 🥺. I read a bit a theory and prototyped the following test workflow:
1. Create two different connections (`client` and `client2`).
2. Open transaction `tx1` in `client`.
3. Read some data in `tx1`.
4. Open transaction `tx2` in `client2`.
5. Modify the same data in `tx2`.
6. Read the same data in `tx1` before commit in `tx2` (test dirty read).
7. Read the same data in `tx1` after commit in `tx2` (test non-repeatable read and phantom read).

Reply:
I also tried to create tests like [here](https://github.com/prisma/prisma/blob/5.8.1/packages/client/tests/functional/batch-transaction-isolation-level/tests.ts#L49-L52), but even set `log_queries=True`, i don't see any transaction SQL queries.
And I don't understand the meaning of test like [here](https://github.com/prisma/prisma/blob/5.8.1/packages/client/tests/functional/interactive-transactions/tests.ts#L710-L725)...

Reply:
Hey sorry for the delay in responding. I'm actually not very familiar with isolation levels myself, I'll look into it.

I've also caused merge conflicts for you (sorry!), I'll fix them.

Reply:
> I also tried to create tests like [here](https://github.com/prisma/prisma/blob/5.8.1/packages/client/tests/functional/batch-transaction-isolation-level/tests.ts?rgh-link-date=2024-01-24T07%3A13%3A25Z#L49-L52), but even set log_queries=True, i don't see any transaction SQL queries.

Yeah unfortunately we don't have a good structured logging setup for the query engine like the JS client which makes things like this really difficult, I suspect they have some separate log level setup somewhere :shrug:

Reply:
@jonathanblade what was the thinking behind creating a second client in the `test_read_uncommited_isolation_level` test?

Reply:
> @jonathanblade what was the thinking behind creating a second client in the `test_read_uncommited_isolation_level` test?

@RobertCraigie To be sure that transactions runs in different connections.

Reply:
> @jonathanblade what was the thinking behind creating a second client in the `test_read_uncommited_isolation_level` test?

Do you think the following is ok?
```python
async with client.tx(isolation_level=prisma.TransactionIsolationLevel.READ_UNCOMMITED) as tx1, client.tx() as tx2:
    # test workflow
```

Reply:
>  To be sure that transactions runs in different connections.

Ah that makes sense, thanks for clarifying 👍

I'm not too sure what's happening with that test case as well, I spent some time looking into it and I can't see why it's not working as expected...

Reply:
Actually, reading the psql docs on this, it looks like they make it difficult to have a dirty read - https://www.postgresql.org/docs/current/transaction-iso.html (only read the start so far).

I'm also seeing the same behaviour in the JS client

```ts
  const user = await prisma.user.create({ data: { name: "First" } });

  await prisma.$transaction(
    async (tx) => {
      await tx.user.update({
        where: { id: user.id },
        data: { name: "Updated" },
      });

      await prisma.$transaction(
        async (tx) => {
          const shouldBeUpdated = await tx.user.findUnique({
            where: { id: user.id },
          });
          console.log("shouldBeUpdated", shouldBeUpdated);
        },
        { isolationLevel: "ReadUncommitted" }
      );
    },
    { isolationLevel: "ReadUncommitted" }
  );
```

Reply:
Hmm... It's very interesting: 

> PostgreSQL's **Read Uncommitted** mode behaves like **Read Committed**. This is because it is the only sensible way to map the standard isolation levels to PostgreSQL's multiversion concurrency control architecture.

Also i found:

> The SQL standard to indicate that dirty read/non-repeatable read/phantom read are **possible** in certain transaction isolation level, but doesn't say they are **required**.

I'll rethink and rewrite the isolation level tests and get back.

Reply:
one way to write the tests could just be to check what the isolation level of the transaction is, if that's even possible. e.g. `tx.query_raw('GET isolation_level')` (made that query up)